packages/nx/src/project-graph/project-graph.ts

```diff
@@ -264,6 +282,47 @@ export async function createProjectGraphAndSourceMapsAsync(
   performance.mark('create-project-graph-async:start');
 
   if (!daemonClient.enabled()) {
+    const lock = !IS_WASM
+      ? new FileLock(join(workspaceDataDirectory, 'project-graph.lock'))
+      : null;
+    let locked = lock?.locked;
+    while (locked) {
+      logger.verbose(
+        'Waiting for graph construction in another process to complete'
+      );
+      const spinner = new DelayedSpinner(
+        'Waiting for graph construction in another process to complete'
+      );
+      await lock.wait();
+      spinner.cleanup();
+
+      // Note: This will currently throw if any of the caches are missing...
+      // It would be nice if one of the processes that was waiting for the lock
+      // could pick up the slack and build the graph if it's missing, but
+      // we wouldn't want either of the below to happen:
+      // - All of the waiting processes to build the graph
+      // - Even one of the processes building the graph on a legitimate error
+
+      const sourceMaps = readSourceMapsCache();
+      if (!sourceMaps) {
+        throw new Error(
+          'The project graph was computed in another process, but the source maps are missing.'
+        );
+      }
+      const graph = await readCachedGraphAndHydrateFileMap();
+      if (graph.errors.length > 0) {
+        throw new ProjectGraphError(graph.errors, graph, sourceMaps);
+      }
+      if (Date.now() - graph.computedAt < 10_000) {
```

Comment:
Get the current time sooner before Nx even starts waiting. The graph will be computing... and should have been written after it has started waiting... This guarantees that it is not old.

Also handle when there is no graph after waiting.